addOns/frontendscanner/src/test/java/org/zaproxy/zap/extensions/frontendscanner/FrontEndScannerProxyListenerTest.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2020 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.frontendscanner;
+
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import org.apache.log4j.Level;
+import org.apache.log4j.Logger;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.parosproxy.paros.model.HistoryReference;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.network.HttpResponseHeader;
+import org.zaproxy.zap.testutils.TestUtils;
+
+/** Unit test for {@link ExtensionFrontEndScannerProxyListener}. */
+public class FrontEndScannerProxyListenerTest extends TestUtils {
+
+    FrontEndScannerProxyListener frontEndScannerProxyListener;
+    HttpMessage msg;
+
+    @BeforeEach
+    public void setUp() {
+        FrontEndScannerAPI api = mock(FrontEndScannerAPI.class);
+        FrontEndScannerOptions options = mock(FrontEndScannerOptions.class);
+        when(options.isEnabled()).thenReturn(true);
+
+        HistoryReference ref = mock(HistoryReference.class);
+        when(ref.getHistoryId()).thenReturn(42);
+
+        Logger rootLogger = Logger.getRootLogger();
+        rootLogger.setLevel(Level.OFF);
+
+        frontEndScannerProxyListener = new FrontEndScannerProxyListener(api, options, rootLogger);
+
+        msg = new HttpMessage();
+        msg.setHistoryRef(ref);
+    }
+
+    @Test
+    public void testInjectTheFrontEndTrackerBeforeOtherScriptsInHeadTag() {
+        // Given
+        String htmlBody = "<head><script></script></head><body></body>";
+        msg.getResponseHeader().setHeader(HttpResponseHeader.CONTENT_TYPE, "text/html");
+        msg.setResponseBody(htmlBody);
+
+        // When
+        frontEndScannerProxyListener.onHttpResponseReceive(msg);
+
+        // Then
+        String expectedHtmlFormat =
+                "<head><script src='https:\\/\\/null\\/zapCallBackUrl\\/-?[0-9]+\\?action=getFile&filename=front-end-scanner.js&historyReferenceId=42'><\\/script><script><\\/script><\\/head><body><\\/body>";
```

Comment:
These are `null` host because there's no request object?

Reply:
It's `null` because the `host` header was not set. I updated the `setUp` so it's more explicit.